src/lib/core/CHIPTLVReader.cpp

```diff
@@ -796,19 +818,25 @@ CHIP_ERROR TLVReader::ReadData(uint8_t * buf, uint32_t len)
     return CHIP_NO_ERROR;
 }
 
-CHIP_ERROR TLVReader::EnsureData(CHIP_ERROR noDataErr)
+CHIP_ERROR TLVReader::EnsureData(CHIP_ERROR noDataErr, bool permitChaining)
 {
     CHIP_ERROR err;
 
     if (mReadPoint == mBufEnd)
     {
+        if (!permitChaining)
```

Comment:
I'm not really sure why this is the right place to put this check.  If we land here, it's almost the _best_ case for GetDataPtr on a segmented backing store: we will get a whole buffer and so forth.

The problem cases are when there is a string or octet string that starts in one buffer and goes on to a different one.  When that happens, GetDataPtr will fail, because EnsureData will succeed but the length will be too big for the current buffer.

If the intent is to make sure we catch use of GetDataPtr on segmented buffers no matter where the buffer boundaries happen to fall for the specific data we are reading, I'd think what we want to do is fail out of GetDataPtr right at the top if `!permitChaining && mBackingStore != nullptr`....

Reply:
You're right, getting to this branch is ideal, because it means that we're reading out the next TLV element on a fresh buffer boundary with no remnants lingering from the previous buffer.

I realize that the buffer straddling case is already being caught earlier in `GetDataPtr`:

```
    // Verify that the entirety of the data is available in the buffer.
    // Note that this may not be possible if the reader is reading from a chain of buffers.
    if (remainingLen < static_cast<uint32_t>(mElemLenOrVal))
        return CHIP_ERROR_TLV_UNDERRUN;
```

That makes my change here un-necessary.

Reply:
I still think it's worth it to fail _all_ attempts to `GetDataPtr` on a reader with multiple buffers.  Otherwise it will sometimes work and sometimes not, depending on exactly where the buffer boundaries fall, which is not a great situation to have to diagnose later...

Reply:
I was under the initial impression that using `TLVBackingStore` doesn't necessarily imply use with chained buffers, because the concrete impl of that interface which is `TLVPacketBufferBackingStore` has a flag to not permit chaining at all. This kind of threw me of, since the description of `TLVBackingStore` fairly definitely implies chaining as the expected requirement of the interface.

Consequently, I was concerned by forcibly erroring out if the `TLVReader` was backed by a `TLVBackingStore` seemed like it might break clients who were using `TLVPacketBufferBackingStore` without chaining enabled.

Scrubbing the code though, it seems like almost all are using `PacketBufferTLVReader`, which is ultimately a `ContiguousBufferTLVReader`, and no one is using `TLVPacketBufferBackingStore`, so I think we should be safe with the more conservative hammer here.

Will make the change.